DIPs/DIP1xxx-JM.md

```diff
@@ -0,0 +1,294 @@
+
+Another solution is to throw type safety out the window, i.e.
+
+```D
+struct DummyType
+{
+    void writelnWithTime(string msg)
+    {
+        File* file = cast(File*)&this;
+        file.writeln(Clock.currTime, " ", msg);
+    }
+}
+
+File file;
+DummyType dummy;
+auto ourDelegate = &dummy.writelnWithTime;
+ourDelegate.ptr = cast(void*)&file;
+```
+
+This also has a cost and I've listed the problems at both the definition site and the call site.
+
+#### At the function definition site:
+
+* requires the function to be defined inside a dummy wrapper type
+* requires a cast of the dummy type context pointer to the actual type you want, this throws away any type safety of the context pointer type and incurs some runtime overhead
+
+#### At the function call site:
+
+* requires an instance of the dummy type to be declared. note that this declaration has no purpose other than to get a delegate to the function
+* once the delegate is retreived, the ptr is set as an independent statement which creates a new place for a runtime error
+* to set the delegate context pointer, the type must be cast to a void*, like the definition site this throws away any type safety for the context pointer type and incurs some runtime overhead
+
+With `extern(delegate)` the solution is simple, clear, and type-safe. No dummy wrapper type, no casting at the definition site or the call site, no runtime overhead, and delegate retrieval happens in one type-safe statement.
+
+```D
+extern(delegate) void writelnWithTime(ref File file, string msg)
+{
+    file.writeln(Clock.currTime, " ", msg);
+}
+
+File file;
+auto ourDelegate = &file.writelnWithTime;
+```
+
+### Breaking changes / deprecation process
+
+No breakage. It requires new syntax that if used previously would have caused a compilation error.
+
+### Examples
+Example to demonstrate the usage:
+```D
+struct Foo
+{
+    void bar(int x)
+    {
+    }
+}
+extern(delegate) void baz(ref Foo foo, int x)
+{
+}
+
+// Note that even though `bar` is defined as a member function inside
+// the struct Foo, the `baz` function has an identical ABI with `bar`.
+
+void delegate(int x) dg;
+Foo foo;
+
+dg = &foo.bar;  // a normal member function delegate
+dg(42);         // calls foo.bar(42)
+
+dg = &foo.baz;  // using UFCS to retreive a delegate to the extern(delegate) baz function
+dg(42);         // calls baz(foo, 42);
+
+dg = &baz;      // a "null delegate", funcptr points to the baz function, but ptr is null.
```

Comment:
I've been thinking about this.  I think I'm going to remove the implicit conversion between `extern(delegate)` function pointers and delegates.  This will prevent the errors I believe you were asking about.  With this change, I don't see any reason to disallow someone from taking the address of an `extern(delegate)` function.  At that point it's behaves just like a normal function pointer.